crates/core/src/host/scheduler.rs

```diff
@@ -249,52 +264,173 @@ impl SchedulerActor {
             }
-            SchedulerMessage::Cancel { id } => {
-                if let Some(key) = self.key_map.remove(&id) {
-                    self.queue.remove(&key);
-                }
-            }
         }
     }
 
     async fn handle_queued(&mut self, id: Expired<ScheduledReducerId>) {
         let id = id.into_inner();
         self.key_map.remove(&id);
-        let Some(module_host) = self.module_host.upgrade() else {
-            return;
-        };
-        let Some(scheduled) = self.db.get(id.0.to_le_bytes()).unwrap() else {
-            return;
-        };
-        let scheduled: ScheduledReducer = bsatn::from_slice(&scheduled).unwrap();
-
-        let db = self.db.clone();
-        tokio::spawn(async move {
-            let info = module_host.info();
-            let identity = info.identity;
-            // TODO: pass a logical "now" timestamp to this reducer call, but there's some
-            //       intricacies to get right (how much drift to tolerate? what kind of tokio::time::MissedTickBehavior do we want?)
-            let res = module_host
-                .call_reducer(
-                    identity,
-                    // Scheduled reducers take `None` as the caller address.
-                    None,
-                    None,
-                    None,
-                    None,
-                    &scheduled.reducer,
-                    ReducerArgs::Bsatn(scheduled.bsatn_args.into()),
-                )
-                .await;
-            if !matches!(res, Err(ReducerCallError::NoSuchModule(_))) {
-                // if we didn't actually call the reducer because the module exited, leave
-                // the ScheduledReducer in the database for when the module restarts
-                let _ = db.remove(id.0.to_le_bytes());
-            }
-            match res {
-                Ok(_) => {}
-                Err(e) => log::error!("invoking scheduled reducer failed: {e:#}"),
+
+        if let Some(module_host) = self.module_host.upgrade() {
+            let db = module_host.dbic().relational_db.clone();
+            let tx = db.begin_mut_tx(IsolationLevel::Serializable);
+            let identity = module_host.info().identity;
+            match self.proccess_schedule(id, &db, &tx) {
+                Ok((reducer, is_repeated)) => {
+                    let _ = tokio::spawn(async move {
+                        let res = module_host
+                            .call_reducer(
+                                Some(tx),
+                                identity,
+                                None,
+                                None,
+                                None,
+                                None,
+                                &reducer.reducer,
+                                ReducerArgs::Bsatn(reducer.bsatn_args.into()),
+                            )
+                            .await;
+                        // Delete the entry from table if it is not repeated schedule in new tx
```

Comment:
You may be aware of this: the check for `Err(NoSuchModule(_))` in the old code is quite important, in case the module host is exited in between the upgrade of the weak reference succeeding and the actual call. We want to re-schedule once the new module host is instantiated.